File: rust/plugins/admob/android/src/main/java/app/nativeblade/admob/AdMobPlugin.kt
package app.nativeblade.admob

import android.app.Activity
import android.content.pm.ApplicationInfo
import android.view.Gravity
import android.view.View
import android.view.ViewGroup
import android.widget.FrameLayout
import app.tauri.annotation.Command
import app.tauri.annotation.InvokeArg
import app.tauri.annotation.TauriPlugin
import app.tauri.plugin.Invoke
import app.tauri.plugin.JSObject
import app.tauri.plugin.Plugin
import com.google.android.gms.ads.AdError
import com.google.android.gms.ads.AdListener
import com.google.android.gms.ads.AdRequest
import com.google.android.gms.ads.AdSize
import com.google.android.gms.ads.AdView
import com.google.android.gms.ads.FullScreenContentCallback
import com.google.android.gms.ads.LoadAdError
import com.google.android.gms.ads.MobileAds
import com.google.android.gms.ads.RequestConfiguration
import com.google.android.gms.ads.interstitial.InterstitialAd
import com.google.android.gms.ads.interstitial.InterstitialAdLoadCallback
import com.google.android.gms.ads.rewarded.RewardedAd
import com.google.android.gms.ads.rewarded.RewardedAdLoadCallback
import com.google.android.ump.ConsentInformation
import com.google.android.ump.ConsentRequestParameters
import com.google.android.ump.UserMessagingPlatform

@InvokeArg
class ConsentArgs {
    var testDeviceIds: List<String> = emptyList()
}

@InvokeArg
class RewardedArgs {
    lateinit var unit: String
    var id: String? = null
}

@InvokeArg
class InterstitialArgs {
    lateinit var unit: String
    var id: String? = null
    var minInterval: Long? = null
}

@InvokeArg
class BannerArgs {
    lateinit var unit: String
    var id: String? = null
}

@TauriPlugin
class AdMobPlugin(private val activity: Activity) : Plugin(activity) {

    companion object {
        // Google's reserved test ad unit ids — served automatically in debug
        // builds so a developer never risks clicking a live ad (account ban).
        private const val TEST_REWARDED = "ca-app-pub-3940256099942544/5224354917"
        private const val TEST_INTERSTITIAL = "ca-app-pub-3940256099942544/1033173712"
        private const val TEST_BANNER = "ca-app-pub-3940256099942544/9214589741"

        // Per-unit last-shown timestamps for interstitial frequency capping.
        private val lastShown = HashMap<String, Long>()
    }

    // Set once test devices are registered: real ad units then serve test ads
    // on those devices, so we stop substituting Google's test unit and exercise
    // the real id safely.
    private var hasTestDevices: Boolean = false

    private var webViewRef: android.webkit.WebView? = null
    private var bannerView: AdView? = null
    private var bannerUnit: String? = null
    private var bannerBuiltWidth = 0
    private var bannerLayoutListener: View.OnLayoutChangeListener? = null

    private val debuggable: Boolean
        get() = (activity.applicationInfo.flags and ApplicationInfo.FLAG_DEBUGGABLE) != 0

    override fun load(webView: android.webkit.WebView) {
        super.load(webView)
        webViewRef = webView
        MobileAds.initialize(activity) { }
    }

    @Command
    fun requestConsent(invoke: Invoke) {
        val args = invoke.parseArgs(ConsentArgs::class.java)

        if (args.testDeviceIds.isNotEmpty()) {
            hasTestDevices = true
            MobileAds.setRequestConfiguration(
                RequestConfiguration.Builder().setTestDeviceIds(args.testDeviceIds).build()
            )
        }

        val params = if (debuggable && args.testDeviceIds.isNotEmpty()) {
            val debugSettings = com.google.android.ump.ConsentDebugSettings.Builder(activity)
                .setDebugGeography(com.google.android.ump.ConsentDebugSettings.DebugGeography.DEBUG_GEOGRAPHY_EEA)
            for (id in args.testDeviceIds) debugSettings.addTestDeviceHashedId(id)
            ConsentRequestParameters.Builder().setConsentDebugSettings(debugSettings.build()).build()
        } else {
            ConsentRequestParameters.Builder().build()
        }

        val consentInformation = UserMessagingPlatform.getConsentInformation(activity)
        consentInformation.requestConsentInfoUpdate(
            activity,
            params,
            {
                UserMessagingPlatform.loadAndShowConsentFormIfRequired(activity) {
                    val result = JSObject()
                    result.put("canRequestAds", consentInformation.canRequestAds())
                    invoke.resolve(result)
                }
            },
            { error ->
                val result = JSObject()
                result.put("canRequestAds", consentInformation.canRequestAds())
                result.put("error", error.message)
                invoke.resolve(result)
            }
        )
    }

    @Command
    fun showRewarded(invoke: Invoke) {
        val args = invoke.parseArgs(RewardedArgs::class.java)
        val unit = if (debuggable && !hasTestDevices) TEST_REWARDED else args.unit

        activity.runOnUiThread {
            RewardedAd.load(activity, unit, AdRequest.Builder().build(), object : RewardedAdLoadCallback() {
                override fun onAdFailedToLoad(error: LoadAdError) {
                    invoke.resolve(failure(error.message))
                }

                override fun onAdLoaded(ad: RewardedAd) {
                    var earned = false
                    var amount = 0
                    var type = ""

                    ad.fullScreenContentCallback = object : FullScreenContentCallback() {
                        override fun onAdFailedToShowFullScreenContent(error: AdError) {
                            invoke.resolve(failure(error.message))
                        }

                        override fun onAdDismissedFullScreenContent() {
                            val result = JSObject()
                            result.put("status", "dismissed")
                            result.put("earned", earned)
                            result.put("amount", amount)
                            result.put("type", type)
                            invoke.resolve(result)
                        }
                    }

                    ad.show(activity) { reward ->
                        earned = true
                        amount = reward.amount
                        type = reward.type
                    }
                }
            })
        }
    }

    @Command
    fun showInterstitial(invoke: Invoke) {
        val args = invoke.parseArgs(InterstitialArgs::class.java)
        val unit = if (debuggable && !hasTestDevices) TEST_INTERSTITIAL else args.unit

        val minInterval = args.minInterval
        if (minInterval != null) {
            val last = lastShown[args.unit] ?: 0L
            if (System.currentTimeMillis() - last < minInterval * 1000) {
                val result = JSObject()
                result.put("status", "capped")
                invoke.resolve(result)
                return
            }
        }

        activity.runOnUiThread {
            InterstitialAd.load(activity, unit, AdRequest.Builder().build(), object : InterstitialAdLoadCallback() {
                override fun onAdFailedToLoad(error: LoadAdError) {
                    invoke.resolve(failure(error.message))
                }

                override fun onAdLoaded(ad: InterstitialAd) {
                    ad.fullScreenContentCallback = object : FullScreenContentCallback() {
                        override fun onAdFailedToShowFullScreenContent(error: AdError) {
                            invoke.resolve(failure(error.message))
                        }

                        override fun onAdDismissedFullScreenContent() {
                            val result = JSObject()
                            result.put("status", "dismissed")
                            invoke.resolve(result)
                        }
                    }
                    lastShown[args.unit] = System.currentTimeMillis()
                    ad.show(activity)
                }
            })
        }
    }

    @Command
    fun showBanner(invoke: Invoke) {
        val args = invoke.parseArgs(BannerArgs::class.java)
        val unit = if (debuggable && !hasTestDevices) TEST_BANNER else args.unit

        activity.runOnUiThread {
            // Showing again replaces the current banner (e.g. a new unit).
            removeBanner()

            if (webViewRef == null) {
                invoke.resolve(failure("no webview"))
                return@runOnUiThread
            }

            bannerUnit = unit
            attachBanner(unit, invoke)
            watchLayoutChanges()
        }
    }

    @Command
    fun hideBanner(invoke: Invoke) {
        activity.runOnUiThread {
            removeBanner()
            invoke.resolve()
        }
    }

    /**
     * UI thread only. Builds an adaptive banner for the current width, anchors
     * it at the bottom and shrinks the WebView to make room. `invoke` is null
     * when rebuilding after a width change, where there is no caller to answer.
     */
    private fun attachBanner(unit: String, invoke: Invoke?) {
        val webView = webViewRef ?: return
        val content = activity.findViewById<ViewGroup>(android.R.id.content)

        val metrics = activity.resources.displayMetrics
        val widthPx = content?.width?.takeIf { it > 0 } ?: metrics.widthPixels
        val adSize = AdSize.getCurrentOrientationAnchoredAdaptiveBannerAdSize(
            activity, (widthPx / metrics.density).toInt()
        )
        val heightPx = adSize.getHeightInPixels(activity)
        // The app is edge-to-edge, so the content view extends under the
        // navigation bar; anchor the banner above it (ads may not be obscured
        // by system bars).
        val navInset = bottomInsetPx()
        bannerBuiltWidth = widthPx

        val banner = AdView(activity)
        banner.adUnitId = unit
        banner.setAdSize(adSize)

        var resolved = false
        banner.adListener = object : AdListener() {
            override fun onAdLoaded() {
                if (resolved) return
                resolved = true
                if (invoke != null) {
                    val result = JSObject()
                    result.put("status", "shown")
                    invoke.resolve(result)
                }
            }

            override fun onAdFailedToLoad(error: LoadAdError) {
                // Refresh failures after the first fill keep the last ad on
                // screen; only tear down when the initial load fails.
                if (resolved) return
                resolved = true
                if (bannerView === banner) removeBanner()
                invoke?.resolve(failure(error.message))
            }
        }

        val params = FrameLayout.LayoutParams(
            FrameLayout.LayoutParams.MATCH_PARENT,
            heightPx,
            Gravity.BOTTOM or Gravity.CENTER_HORIZONTAL
        )
        params.bottomMargin = navInset
        activity.addContentView(banner, params)

        // Reserve the space up front so the page lays out once, not again
        // when the ad fills.
        (webView.layoutParams as? ViewGroup.MarginLayoutParams)?.let {
            it.bottomMargin = navInset + heightPx
            webView.layoutParams = it
        }

        bannerView = banner
        banner.loadAd(AdRequest.Builder().build())
    }

    /**
     * Rebuild the banner when the usable width changes (rotation, resize) —
     * an adaptive banner is sized for the width it was loaded with. The
     * activity is not recreated on rotation (Tauri uses configChanges), so
     * this is the only signal. Registered once; a no-op while no banner shows.
     */
    private fun watchLayoutChanges() {
        if (bannerLayoutListener != null) return
        val content = activity.findViewById<ViewGroup>(android.R.id.content) ?: return

        val listener = View.OnLayoutChangeListener { v, left, _, right, _, _, _, _, _ ->
            val width = right - left
            if (bannerUnit == null || width == 0 || width == bannerBuiltWidth) {
                return@OnLayoutChangeListener
            }
            // Rebuilding mutates the hierarchy; defer until the layout pass ends.
            v.post {
                val unit = bannerUnit ?: return@post
                if (v.width == 0 || v.width == bannerBuiltWidth) return@post
                detachBanner()
                attachBanner(unit, null)
            }
        }
        bannerLayoutListener = listener
        content.addOnLayoutChangeListener(listener)
    }

    /** UI thread only. Removes the banner view and gives the WebView its space back. */
    private fun detachBanner() {
        bannerView?.let { banner ->
            (banner.parent as? ViewGroup)?.removeView(banner)
            banner.destroy()
        }
        bannerView = null
        bannerBuiltWidth = 0

        val webView = webViewRef ?: return
        (webView.layoutParams as? ViewGroup.MarginLayoutParams)?.let {
            if (it.bottomMargin != 0) {
                it.bottomMargin = 0
                webView.layoutParams = it
            }
        }
    }

    /** UI thread only. Fully stops the banner, including rebuild-on-rotation. */
    private fun removeBanner() {
        bannerUnit = null
        detachBanner()
    }

    private fun bottomInsetPx(): Int {
        val insets = activity.window.decorView.rootWindowInsets ?: return 0
        return if (android.os.Build.VERSION.SDK_INT >= 30) {
            insets.getInsets(android.view.WindowInsets.Type.navigationBars()).bottom
        } else {
            @Suppress("DEPRECATION")
            insets.systemWindowInsetBottom
        }
    }

    private fun failure(message: String?): JSObject {
        val result = JSObject()
        result.put("status", "failed")
        result.put("error", message ?: "ad failed")
        return result
    }
}

File: rust/plugins/admob/ios/Sources/AdMobPlugin.swift
import Foundation
#if canImport(GoogleMobileAds)
import GoogleMobileAds
import UserMessagingPlatform
import AppTrackingTransparency
import UIKit
#endif
import SwiftRs
import Tauri

struct NBConsentArgs: Decodable {
    let testDeviceIds: [String]?
}

struct NBRewardedArgs: Decodable {
    let unit: String
    let id: String?
}

struct NBInterstitialArgs: Decodable {
    let unit: String
    let id: String?
    let minInterval: Double?
}

struct NBBannerArgs: Decodable {
    let unit: String
    let id: String?
}

class AdMobPlugin: Plugin {
    // Google's reserved test ad unit ids, served in DEBUG builds so a developer
    // never risks clicking a live ad (account ban).
    private static let testRewarded = "ca-app-pub-3940256099942544/1712485313"
    private static let testInterstitial = "ca-app-pub-3940256099942544/4411468910"
    private static let testBanner = "ca-app-pub-3940256099942544/2435281174"

    private static var lastShown: [String: TimeInterval] = [:]

    // Set once test devices are registered: real ad units then serve test ads
    // on those devices, so we stop substituting the test unit and exercise the
    // real id safely.
    private static var hasTestDevices = false

    #if canImport(GoogleMobileAds)
    private var rewardedAd: RewardedAd?
    private var interstitialAd: InterstitialAd?
    private var contentDelegate: FullScreenDelegate?
    private var bannerView: BannerView?
    private var bannerDelegate: BannerAdDelegate?
    private var bannerUnit: String?
    private var bannerBuiltWidth: CGFloat = 0
    private var orientationObserver: NSObjectProtocol?
    // Callers waiting on the in-flight consent flow. Main thread only.
    private var consentWaiters: [Invoke] = []
    #endif
    private weak var webviewRef: WKWebView?

    @objc public override func load(webview: WKWebView) {
        super.load(webview: webview)
        webviewRef = webview
        #if canImport(GoogleMobileAds)
        MobileAds.shared.start(completionHandler: nil)
        #endif
    }

    @objc public func requestConsent(_ invoke: Invoke) {
        #if canImport(GoogleMobileAds)
        let testIds = (try? invoke.parseArgs(NBConsentArgs.self))?.testDeviceIds ?? []

        DispatchQueue.main.async {
            // Test device registration is a global SDK side effect that every
            // caller's later ad loads depend on, so apply it even for calls
            // that coalesce below.
            if !testIds.isEmpty {
                Self.hasTestDevices = true
                MobileAds.shared.requestConfiguration.testDeviceIdentifiers = testIds
            }

            // The ATT prompt and UMP form are modal and global: a second flow
            // started while one is up queues another form the user must
            // dismiss again. Coalesce concurrent calls into the in-flight flow.
            self.consentWaiters.append(invoke)
            if self.consentWaiters.count > 1 { return }

            ATTrackingManager.requestTrackingAuthorization { _ in
                DispatchQueue.main.async {
                    let params = RequestParameters()
                    // Force the EEA consent form in debug for registered test
                    // devices, mirroring the Android side.
                    if Self.isDebug && !testIds.isEmpty {
                        let debugSettings = DebugSettings()
                        debugSettings.geography = .EEA
                        debugSettings.testDeviceIdentifiers = testIds
                        params.debugSettings = debugSettings
                    }
                    ConsentInformation.shared.requestConsentInfoUpdate(with: params) { error in
                        let finish: () -> Void = {
                            self.finishConsent(ConsentInformation.shared.canRequestAds, error?.localizedDescription)
                        }
                        if let root = Self.rootViewController() {
                            ConsentForm.loadAndPresentIfRequired(from: root) { _ in finish() }
                        } else {
                            finish()
                        }
                    }
                }
            }
        }
        #else
        invoke.resolve(["canRequestAds": false])
        #endif
    }

    #if canImport(GoogleMobileAds)
    /// Main thread only. Resolves every caller waiting on the shared consent flow.
    private func finishConsent(_ canRequestAds: Bool, _ error: String?) {
        let waiters = consentWaiters
        consentWaiters.removeAll()
        for invoke in waiters {
            invoke.resolve([
                "canRequestAds": canRequestAds,
                "error": error as Any,
            ])
        }
    }
    #endif

    @objc public func showRewarded(_ invoke: Invoke) {
        #if canImport(GoogleMobileAds)
        guard let args = try? invoke.parseArgs(NBRewardedArgs.self) else {
            invoke.resolve(Self.failure("invalid args"))
            return
        }
        let unit = (Self.isDebug && !Self.hasTestDevices) ? Self.testRewarded : args.unit

        RewardedAd.load(with: unit, request: Request()) { [weak self] ad, error in
            guard let self = self else { return }
            if let error = error {
                invoke.resolve(Self.failure(error.localizedDescription))
                return
            }
            guard let ad = ad, let root = Self.rootViewController() else {
                invoke.resolve(Self.failure("no ad or root view controller"))
                return
            }

            self.rewardedAd = ad
            var earned = false
            let delegate = FullScreenDelegate { reason in
                let reward = ad.adReward
                invoke.resolve([
                    "status": reason,
                    "earned": earned,
                    "amount": reward.amount,
                    "type": reward.type,
                ])
            }
            self.contentDelegate = delegate
            ad.fullScreenContentDelegate = delegate

            // The present handler is the reward callback: it fires only when the
            // user actually earns the reward (mirrors the Android show listener).
            // It runs before the dismiss delegate, so `earned` is set by then.
            ad.present(from: root) {
                earned = true
            }
        }
        #else
        invoke.resolve(Self.failure("unsupported"))
        #endif
    }

    @objc public func showInterstitial(_ invoke: Invoke) {
        #if canImport(GoogleMobileAds)
        guard let args = try? invoke.parseArgs(NBInterstitialArgs.self) else {
            invoke.resolve(Self.failure("invalid args"))
            return
        }

        if let minInterval = args.minInterval {
            let last = Self.lastShown[args.unit] ?? 0
            if Date().timeIntervalSince1970 - last < minInterval {
                invoke.resolve(["status": "capped"])
                return
            }
        }

        let unit = (Self.isDebug && !Self.hasTestDevices) ? Self.testInterstitial : args.unit

        InterstitialAd.load(with: unit, request: Request()) { [weak self] ad, error in
            guard let self = self else { return }
            if let error = error {
                invoke.resolve(Self.failure(error.localizedDescription))
                return
            }
            guard let ad = ad, let root = Self.rootViewController() else {
                invoke.resolve(Self.failure("no ad or root view controller"))
                return
            }

            self.interstitialAd = ad
            let delegate = FullScreenDelegate { reason in
                invoke.resolve(["status": reason])
            }
            self.contentDelegate = delegate
            ad.fullScreenContentDelegate = delegate

            Self.lastShown[args.unit] = Date().timeIntervalSince1970
            ad.present(from: root)
        }
        #else
        invoke.resolve(Self.failure("unsupported"))
        #endif
    }

    @objc public func showBanner(_ invoke: Invoke) {
        #if canImport(GoogleMobileAds)
        guard let args = try? invoke.parseArgs(NBBannerArgs.self) else {
            invoke.resolve(Self.failure("invalid args"))
            return
        }
        let unit = (Self.isDebug && !Self.hasTestDevices) ? Self.testBanner : args.unit

        DispatchQueue.main.async {
            // Showing again replaces the current banner (e.g. a new unit).
            self.removeBanner()

            guard self.webviewRef?.superview != nil else {
                invoke.resolve(Self.failure("no webview"))
                return
            }

            self.bannerUnit = unit
            self.attachBanner(unit: unit, invoke: invoke)
            self.watchLayoutChanges()
        }
        #else
        invoke.resolve(Self.failure("unsupported"))
        #endif
    }

    @objc public func hideBanner(_ invoke: Invoke) {
        #if canImport(GoogleMobileAds)
        DispatchQueue.main.async {
            self.removeBanner()
            invoke.resolve()
        }
        #else
        invoke.resolve()
        #endif
    }

    #if canImport(GoogleMobileAds)
    /// Main thread only. Builds an adaptive banner for the current width,
    /// anchors it above the home indicator and shrinks the WebView to make
    /// room. `invoke` is nil when rebuilding after a rotation, where there is
    /// no caller to answer.
    private func attachBanner(unit: String, invoke: Invoke?) {
        guard let webview = webviewRef, let superview = webview.superview,
              let root = Self.rootViewController() else {
            invoke?.resolve(Self.failure("no webview"))
            return
        }

        let bounds = superview.bounds
        // Keep the banner above the home indicator (ads may not sit under
        // system UI).
        let bottomInset = superview.safeAreaInsets.bottom
        let adSize = currentOrientationAnchoredAdaptiveBanner(width: bounds.width)
        let height = adSize.size.height
        bannerBuiltWidth = bounds.width

        let banner = BannerView(adSize: adSize)
        banner.adUnitID = unit
        banner.rootViewController = root
        banner.frame = CGRect(
            x: 0,
            y: bounds.height - bottomInset - height,
            width: bounds.width,
            height: height
        )
        banner.autoresizingMask = [.flexibleWidth, .flexibleTopMargin]

        let delegate = BannerAdDelegate(
            onLoad: {
                invoke?.resolve(["status": "shown"])
            },
            onFail: { [weak self] message in
                if let self = self, self.bannerView === banner {
                    self.removeBanner()
                }
                invoke?.resolve(Self.failure(message))
            }
        )
        bannerDelegate = delegate
        banner.delegate = delegate

        superview.addSubview(banner)
        bannerView = banner

        // Reserve the space up front so the page lays out once, not again
        // when the ad fills.
        var frame = webview.frame
        frame.size.height = banner.frame.minY - frame.origin.y
        webview.frame = frame

        banner.load(Request())
    }

    /// Rebuild the banner when the usable width changes (rotation) — an
    /// adaptive banner is sized for the width it was loaded with. Registered
    /// while a banner is showing; removeBanner() unregisters it.
    private func watchLayoutChanges() {
        if orientationObserver != nil { return }
        UIDevice.current.beginGeneratingDeviceOrientationNotifications()
        orientationObserver = NotificationCenter.default.addObserver(
            forName: UIDevice.orientationDidChangeNotification,
            object: nil,
            queue: .main
        ) { [weak self] _ in
            // The notification fires before the window finishes rotating; give
            // the layout pass time to settle before measuring the new width.
            DispatchQueue.main.asyncAfter(deadline: .now() + 0.15) {
                guard let self = self, let unit = self.bannerUnit,
                      let width = self.webviewRef?.superview?.bounds.width,
                      width != self.bannerBuiltWidth else { return }
                self.detachBanner()
                self.attachBanner(unit: unit, invoke: nil)
            }
        }
    }

    /// Main thread only. Removes the banner view and gives the WebView its
    /// space back.
    private func detachBanner() {
        bannerView?.removeFromSuperview()
        bannerView = nil
        bannerDelegate = nil
        bannerBuiltWidth = 0

        if let webview = webviewRef, let superview = webview.superview {
            webview.frame = superview.bounds
        }
    }

    /// Main thread only. Fully stops the banner, including rebuild-on-rotation.
    private func removeBanner() {
        bannerUnit = nil
        detachBanner()

        // Stop watching for rotation: device-orientation notifications keep
        // the accelerometer reporting active, which costs battery for nothing
        // once no banner is showing.
        if let observer = orientationObserver {
            NotificationCenter.default.removeObserver(observer)
            orientationObserver = nil
            UIDevice.current.endGeneratingDeviceOrientationNotifications()
        }
    }
    #endif

    private static var isDebug: Bool {
        #if DEBUG
        return true
        #else
        return false
        #endif
    }

    private static func failure(_ message: String) -> [String: Any] {
        return ["status": "failed", "error": message]
    }

    #if canImport(GoogleMobileAds)
    private static func rootViewController() -> UIViewController? {
        let scenes = UIApplication.shared.connectedScenes.compactMap { $0 as? UIWindowScene }
        for scene in scenes {
            if let root = scene.windows.first(where: { $0.isKeyWindow })?.rootViewController {
                return root
            }
        }
        return scenes.first?.windows.first?.rootViewController
    }
    #endif
}

#if canImport(GoogleMobileAds)
class FullScreenDelegate: NSObject, FullScreenContentDelegate {
    private let onFinish: (String) -> Void
    private var finished = false

    init(_ onFinish: @escaping (String) -> Void) {
        self.onFinish = onFinish
    }

    private func finish(_ reason: String) {
        if finished { return }
        finished = true
        onFinish(reason)
    }

    func ad(_ ad: FullScreenPresentingAd, didFailToPresentFullScreenContentWithError error: Error) {
        finish("failed")
    }

    func adDidDismissFullScreenContent(_ ad: FullScreenPresentingAd) {
        finish("dismissed")
    }
}

class BannerAdDelegate: NSObject, BannerViewDelegate {
    private let onLoad: () -> Void
    private let onFail: (String) -> Void
    private var finished = false

    init(onLoad: @escaping () -> Void, onFail: @escaping (String) -> Void) {
        self.onLoad = onLoad
        self.onFail = onFail
    }

    func bannerViewDidReceiveAd(_ bannerView: BannerView) {
        if finished { return }
        finished = true
        onLoad()
    }

    func bannerView(_ bannerView: BannerView, didFailToReceiveAdWithError error: Error) {
        // Refresh failures after the first fill keep the last ad on screen;
        // only report when the initial load fails.
        if finished { return }
        finished = true
        onFail(error.localizedDescription)
    }
}
#endif

@_cdecl("init_plugin_nativeblade_admob")
func initPlugin() -> Plugin {
    return AdMobPlugin()
}
